File: unityProduct/Assets/Script/APIController.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using Newtonsoft.Json.Linq;
using System;

public class APIController
{
    public void GetFgoGetRole(JObject input, Action<APIResult> callBack)
    {
        if (Config.IS_SQLITE)
        {
            SqliteGetRole sqliteGetRole = new SqliteGetRole();
            sqliteGetRole.Send(input, callBack);
        }
        else
        {
            sendAPI("role", input, callBack);
        }
    }

    public void GetFgoRoleResource(JObject input, Action<APIResult> callBack)
    {
        if (Config.IS_SQLITE)
        {
            SqliteGetRoleResource sqliteGetRoleResource = new SqliteGetRoleResource();
            sqliteGetRoleResource.Send(input,callBack);
        }
        else
        {
            sendAPI("role_resource", input, callBack);
        }
    }

    public void AskAiTalk(JObject input, Action<APIResult> callBack)
    {
        if (Config.IS_SQLITE)
        {
            ClientAiTalk clientAiTalk = new ClientAiTalk();
            clientAiTalk.Send(input, callBack);
        }
        else
        {
            postAPI("talk_group_ask", input, callBack);
        }
    }


    private void sendAPI(string url,JObject jobj,Action<APIResult> callBack)
    {
        GameObject g = new GameObject(url);
        UnityEngine.Object.DontDestroyOnLoad(g);
        APIClient client = g.AddComponent<APIClient>();
        client.SendAPI(Config.API_DOMAIN + url, jobj, callBack);
    }

    private void postAPI(string url, JObject jobj, Action<APIResult> callBack)
    {
        GameObject g = new GameObject(url);
        UnityEngine.Object.DontDestroyOnLoad(g);
        APIClient client = g.AddComponent<APIClient>();
        client.SendPOST(Config.API_DOMAIN + url, jobj, callBack);
    }
}

public interface APIResult
{    
    JObject getData();
}

File: FgoProductCocos/assets/script/APIController.ts
import { Config } from "./Config";

export class APIController 
{
    public GetFgoGetRole(input : object, callBack : (result : APIResult) => void)
    {
        this.sendAPI("role", input, callBack);
    }

    public GetFgoRoleResource(input : object, callBack : (result : APIResult) => void)
    {
        this.sendAPI("role_resource", input, callBack);
    }

    public AskAiTalk(input : object, callBack : (result : APIResult) => void)
    {
        this.postAPI("talk_group_ask", input, callBack);
    }

    private sendAPI(url : string,data : object,callback : (result : APIResult) => void)
    {
        var params  : string = "";
        for(var key in data)
        {
            if(params .length == 0)
            {
                params  += "?";
            }
            else
            {
                params  += "&";
            }
            params  += key + "=" + data[key];
        }

        let path : string = Config.API_DOMAIN + url + params;
        fetch(path)
            .then(response => response.json())
            .then(data => {
                console.log('收到資料:', data);
                callback(new DataResult(data));
            })
            .catch(error => {
                console.error('API 錯誤:', error);
                callback(new DataResult(error));
            });
    }

    private postAPI(url : string,data : object,callback : (result : APIResult) => void)
    {
        let path : string = Config.API_DOMAIN + url;

        fetch(path, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(data)
        })
        .then(res => res.json())
        .then(data => {
            console.log('收到資料:', data);
            callback(new DataResult(data));
        })
        .catch(error => {
            console.error('API 錯誤:', error);
            callback(new DataResult(error));
        });
    }

}

export interface APIResult
{
    getData() : object
}

class DataResult implements APIResult
{
    public data : object;
    constructor(data : object)
    {
        this.data = data;
    }
    public getData(): object {
        return this.data;
    }
}